### ISSUE TITLE
Credhub set using CLI > 2.0.0 fails to save values to server 2.0.2

### ISSUE THREAD

**Issue by xtreme-conor-nosal** [cloudfoundry-incubator/credhub-cli#65]
What version of the credhub server you are using?
2.0.2

What version of the credhub cli you are using?
2.0.1
2.1.0

If you were attempting to accomplish a task, what was it you were attempting to do?
Save a value
`credhub s -n /foo -t value -v bar`

What did you expect to happen?
Save the value

What was the actual behavior?
CLI 2.0.1 and 2.1.0 send invalid parameter "mode". CLI 2.0.0 writes successfully.
```
[DEBUG] PUT /api/v1/data HTTP/1.1
Host: 10.0.0.6:8844
User-Agent: Go-http-client/1.1
Content-Length: 63
Content-Type: application/json
Accept-Encoding: gzip

{"mode":"overwrite","name":"/foo","type":"value","value":"bar"}
[DEBUG] HTTP/1.1 422
Transfer-Encoding: chunked
Cache-Control: no-cache, no-store, max-age=0, must-revalidate
Content-Type: application/json;charset=UTF-8
Date: Wed, 14 Nov 2018 19:06:00 GMT
Expires: 0
Pragma: no-cache
Strict-Transport-Security: max-age=31536000 ; includeSubDomains
X-Content-Type-Options: nosniff
X-Frame-Options: DENY
X-Xss-Protection: 1; mode=block

81
{"error":"The request includes an unrecognized parameter 'mode'. Please update or remove this parameter and retry your request."}
0
```

**Comment by mdelillo** [comment]
Hey @xtreme-conor-nosal, we're trying to reproduce this and could use some more information.

1. Have you upgraded either the credhub server or CLI before this started happening?
1. What is the `ServerVersion` field in `~/.credhub/config.json`?
1. What is your workflow for targeting/logging into CredHub? Do you set environment variables? Do you use the `credhub login` or `credhub api` commands?

**Comment by xtreme-conor-nosal** [comment]
1. The CLI was updated on the local workstation. The server is deployed as part of a bosh-bootloader deployment. We don't upgrade in place - we always repave.
2. The `ServerVersion` in `~/.credhub/config.json` is 1.6.5
3. We set environment variables:
```
CREDHUB_SERVER=https://10.0.0.6:8844
CREDHUB_CA_CERT=$CERT
CREDHUB_PROXY=ssh+socks5://jumpbox@$IP:22?private-key=/tmp/bbl_cWbdY8
CREDHUB_CLIENT=credhub-admin
CREDHUB_SECRET=$SECRET
```

**Comment by mdelillo** [comment]
The issue seems to be that your `~/.credhub/config.json` says that the server version is `1.6.5`. We introduced caching of this in CLI version `2.0.1` which is why `2.0.0` still works for you.

You can fix this issue by running `credhub login` using the latest CLI after setting those environment variables. This is in fact how the CLI should be used (although I recognize that it is not well documented). That will update the `ServerVersion` in the config file as well as update your UAA auth token.

**Comment by xtreme-conor-nosal** [comment]
We use the environment variables to avoid writing the client secret to disk, and to try and avoid stale properties. 

For consistency, would it be possible to control the `ServerVersion` via the environment?

e.g. one of:
```
CREDHUB_SERVER_VERSION=2.0.2
CREDHUB_API_VERSION=2
CREDHUB_DISABLE_VERSION_CACHE=true
```

As a workaround, would something like `curl /version > config.json` work in combination with the environment variables?

**Comment by TisVictress** [comment]
@xtreme-conor-nosal The cli relies on `~/.credhub/config.json` to cache the server version.  Unfortunately, this file will be overwritten by the cli when you call `credhub login`.  If you try to replace values in the config.json file with environment variables, then they will be overwritten the next time you call `credhub login`.  Does that help?

**Comment by mdelillo** [comment]
@xtreme-conor-nosal we've had some discussion about using only environment variables when running the CLI (https://github.com/cloudfoundry-incubator/credhub-cli/issues/34), but it would be a pretty large change and probably not something we'll get around to in the near future.

Your workaround of modifying the config file will definitely work, but I wouldn't suggest it as a long-term solution. I understand your concern of not wanting to write the client secret to disk, and using `credhub login` wouldn't actually require that. If you run `credhub login` with no arguments, it will use the `CREDHUB_*` environment variables. But it will also ensure that the server version is correct by making a request to credhub.

**Comment by mdelillo** [comment]
@xtreme-conor-nosal We just tried it out and you're right, the UAA creds get written to the config file even when using environment variables. We made a story [here](https://www.pivotaltracker.com/story/show/162201515) to stop writing them to disk.

**Comment by ankeesler** [comment]
Fixed in [v2.2.1](https://github.com/cloudfoundry-incubator/credhub-cli/releases/tag/2.2.1).
